This is a guided project on Dataquest for exploring Ebay Car Sales Data. In this project, I'll be cleaning the dataset.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In the dataset, the fields [vehicleType, gearbox, model, fuelType, notRepairedDamage] have null entries, but none have more than 20% null values. The column names use camelcase instead of Python's preferred snakecase, so next I will be converting the columns names to snakecase. 

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage','ad_created', 'number_of_pictures','postal_code','last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The seller column has 49999/50000 entries = privat. The offer_type column has 49999/50000 entries = Angebot. These 2 columns can be dropped.
The odometer and price columns are stored as text fields because they include characters ('$' and 'km'). 

In [6]:
autos.drop(['seller','offer_type'], axis=1)
autos["price"] = autos["price"].str.replace('$','').str.replace(',','')
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].str.replace('km','').str.replace(',','')
autos["odometer"] = autos["odometer"].astype(int)
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  number_of_pictures  \
count              40171                50000             50000.0   
unique                 2                   76                 NaN   
top                 nein  2016-04-03 00:00:00                 NaN   
freq               35232                 1946                 NaN   
mean                 NaN                  NaN                 0.0   
std                  NaN                  NaN                 0.0   
min                  NaN                  NaN                 0.0   
25%                  NaN                  NaN                 0.0   
50%                  NaN                  NaN                 0.0   
75%                  NaN                  NaN                 0.0   
max                  NaN                  NaN                 0.0   

         postal_code            last_seen  
count   50000.000000                50000  
unique           NaN                39481  
top              NaN  2016-04-07 06:17:27  
freq             NaN                    8  
mean    50813.627300                  NaN  
std     25779.747957                  NaN  
min      1067.000000                  NaN  
25%     30451.000000                  NaN  
50%     49577.000000                  NaN  
75%     71540.000000                  NaN  
max     99998.000000   

Next, I will look into any outliers in the odometer_km and price columns

In [7]:
autos["odometer_km"].unique().shape

(13,)

In [8]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [9]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

It appears that the max odometer reading is 150,000 km and 32,424 cars have the max reading.

In [10]:
autos["price"].unique().shape

(2357,)

In [11]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

The minimum price is $0 and the maximum is $100,000,000. These should likely be removed.

In [12]:
autos["price"].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

In [13]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

I've decided to remove rows where the price is less than 1000 or more than 4,000,000, arbitrarily. 

In [14]:
autos = autos[autos["price"].between(1000,4000000)]
autos["price"].describe()

count    3.863200e+04
mean     7.498207e+03
std      2.535787e+04
min      1.000000e+03
25%      2.200000e+03
50%      4.350000e+03
75%      8.950000e+03
max      3.890000e+06
Name: price, dtype: float64

Next, I'll extract just the date values from the date_crawled, ad_created, and last_seen columns

In [15]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025549
2016-03-06    0.013875
2016-03-07    0.035126
2016-03-08    0.032615
2016-03-09    0.032460
2016-03-10    0.033314
2016-03-11    0.032797
2016-03-12    0.037378
2016-03-13    0.015997
2016-03-14    0.036628
2016-03-15    0.033625
2016-03-16    0.029069
2016-03-17    0.030493
2016-03-18    0.012839
2016-03-19    0.035126
2016-03-20    0.038155
2016-03-21    0.037301
2016-03-22    0.032538
2016-03-23    0.032201
2016-03-24    0.029017
2016-03-25    0.030519
2016-03-26    0.033107
2016-03-27    0.031399
2016-03-28    0.035359
2016-03-29    0.033987
2016-03-30    0.033055
2016-03-31    0.031399
2016-04-01    0.034609
2016-04-02    0.036291
2016-04-03    0.039139
2016-04-04    0.036912
2016-04-05    0.013357
2016-04-06    0.003262
2016-04-07    0.001501
Name: date_crawled, dtype: float64

date_crawled ranged from 2016-03-05 to 2016-04-07

In [16]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
2016-01-03    0.000026
2016-01-07    0.000026
2016-01-10    0.000052
2016-01-13    0.000026
2016-01-14    0.000026
2016-01-16    0.000026
2016-01-22    0.000026
2016-01-27    0.000078
2016-01-29    0.000026
2016-02-01    0.000026
2016-02-02    0.000052
2016-02-05    0.000052
2016-02-07    0.000026
2016-02-09    0.000026
2016-02-11    0.000026
2016-02-12    0.000052
2016-02-14    0.000052
2016-02-16    0.000026
2016-02-17    0.000026
2016-02-18    0.000052
2016-02-19    0.000078
2016-02-20    0.000026
2016-02-21    0.000052
2016-02-22    0.000026
2016-02-23    0.000104
                ...   
2016-03-09    0.032641
2016-03-10    0.032978
2016-03-11    0.033055
2016-03-12    0.037119
2016-03-13    0.017654
2016-03-14    0.034971
2016-03-15    0.033444
2016-03-16    0.029639
2016-03-17    0.030182
2016-03-18    0.013279
2016-03-19    0.034065
2016-03-20    0.038258
2016-03-21 

ad_created ranges from 2015-06-11 to 2016-04-07

In [17]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001087
2016-03-06    0.003572
2016-03-07    0.004556
2016-03-08    0.006238
2016-03-09    0.008905
2016-03-10    0.009811
2016-03-11    0.011726
2016-03-12    0.022184
2016-03-13    0.008387
2016-03-14    0.011985
2016-03-15    0.014988
2016-03-16    0.015454
2016-03-17    0.026377
2016-03-18    0.007377
2016-03-19    0.014599
2016-03-20    0.019802
2016-03-21    0.019673
2016-03-22    0.020786
2016-03-23    0.017913
2016-03-24    0.018534
2016-03-25    0.017757
2016-03-26    0.016075
2016-03-27    0.014082
2016-03-28    0.019440
2016-03-29    0.020786
2016-03-30    0.023452
2016-03-31    0.022727
2016-04-01    0.023193
2016-04-02    0.024928
2016-04-03    0.024436
2016-04-04    0.023374
2016-04-05    0.131161
2016-04-06    0.234676
2016-04-07    0.139962
Name: last_seen, dtype: float64

last_seen has the same date ranged as date_crawled

In [18]:
autos["registration_year"].describe()

count    38632.000000
mean      2005.678194
std         86.678597
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

There are unexpected values in the registration_year field. Cars did not exist in 1000 and the data was collected in 2016, so the registration year should not be greater than 2016

In [19]:
temp = autos[autos["registration_year"].between(1900,2016)]
temp["registration_year"].describe()

count    37212.000000
mean      2003.967618
std          7.021028
min       1927.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [20]:
temp["registration_year"].value_counts().sort_index().head()

1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64

The oldest car in the dataset has a resitration year of 1927.

In [21]:
autos = autos[autos["registration_year"].between(1927,2016)]

In [22]:
autos["brand"].describe()

count          37212
unique            40
top       volkswagen
freq            7845
Name: brand, dtype: object

In [23]:
top20brands_counts = autos["brand"].value_counts().sort_values(ascending=False).head(20)
print(top20brands_counts)
top20brands = top20brands_counts.index
print(top20brands)

volkswagen        7845
bmw               4665
mercedes_benz     4152
audi              3631
opel              3314
ford              2185
renault           1387
peugeot           1038
fiat               784
skoda              709
seat               643
smart              618
toyota             544
mazda              530
citroen            517
nissan             507
mini               405
hyundai            400
sonstige_autos     391
volvo              334
Name: brand, dtype: int64
Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda', 'seat', 'smart', 'toyota', 'mazda',
       'citroen', 'nissan', 'mini', 'hyundai', 'sonstige_autos', 'volvo'],
      dtype='object')


I've decided to aggregate data for the top 20 most listed brands of cars in the dataset. For each of these brands, I'll find the mean price of car. 

In [24]:
autos.loc[autos["brand"]=="bmw","price"].mean()

9381.89174705252

In [25]:
mean_price_by_brand = {}
for x in top20brands:
    mean_price_by_brand[x] = autos.loc[autos["brand"]==x,"price"].mean()
    
print(mean_price_by_brand)

{'mini': 10715.237037037037, 'ford': 5786.703432494279, 'citroen': 4614.970986460348, 'hyundai': 6181.8725, 'smart': 3780.4692556634304, 'audi': 10322.269347287249, 'bmw': 9381.89174705252, 'skoda': 6836.696755994359, 'mazda': 5309.526415094339, 'opel': 4219.954737477368, 'toyota': 5573.57169117647, 'fiat': 4008.174744897959, 'renault': 3590.942321557318, 'nissan': 6428.428007889546, 'volvo': 6151.583832335329, 'sonstige_autos': 27654.273657289003, 'volkswagen': 6898.376545570427, 'mercedes_benz': 9302.614402697494, 'seat': 5638.640746500778, 'peugeot': 3955.169556840077}


In [28]:
brand_mean_mileage = {}
for x in top20brands:
    brand_mean_mileage[x] = autos.loc[autos["brand"]==x,"odometer_km"].mean()
bmp_series = pd.Series(mean_price_by_brand)
bmm_series = pd.Series(brand_mean_mileage)
brand_df = pd.DataFrame(bmp_series, columns=['mean_price'])
brand_df['mean_mileage'] = bmm_series
print(brand_df)

                  mean_price   mean_mileage
audi            10322.269347  127491.049298
bmw              9381.891747  132005.359057
citroen          4614.970986  114245.647969
fiat             4008.174745  107901.785714
ford             5786.703432  119622.425629
hyundai          6181.872500  101862.500000
mazda            5309.526415  119981.132075
mercedes_benz    9302.614403  130062.620424
mini            10715.237037   88679.012346
nissan           6428.428008  110335.305720
opel             4219.954737  123952.926976
peugeot          3955.169557  122341.040462
renault          3590.942322  121423.936554
seat             5638.640747  116104.199067
skoda            6836.696756  110063.469676
smart            3780.469256   97775.080906
sonstige_autos  27654.273657   89411.764706
toyota           5573.571691  113740.808824
volkswagen       6898.376546  125771.829191
volvo            6151.583832  137230.538922
